# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](http://nbviewer.jupyter.org/github/jupyter/notebook/blob/master/docs/source/examples/Notebook/Examples%20and%20Tutorials%20Index.ipynb).  Additionally, in order to write text in this document, you will be making use of markdown cells. A handy cheatsheet for markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Chapter 1 Notes

#### 1.1
1. Python is a high level programming language
2. Uses the prompt to execute lines of code or scripts
3. Programming in the promt is called interactive mode and programming scripts is called script mode
4. Interpreters read and execute programs to produce output or whatever the code is telling the computer to do
5. Compilers translate the code before the program executes so that the program can be understood by the programming application

#### 1.3
5. A program can have several types of errors: syntax, runtime, and semantic
6. Syntax errors refer to improper structure of programming language
7. Runtime errors refer to errors that occur while the program is running
8. Semantic errors are errors that do not prevent the program from running but should be fixed

#### 1.4
9. Programming is a formal language that uses tokens (words, numbers, chemical elements) and structure (the way tokens are arranged)
10. Formal languages are much more concise and less redundant than everyday language

#### 1.6
11. Debugging can be very frustrating
12. Patience is a virtue

### Exercise 3

Type `help()` to start the online help utility. Or you can type help('print') to get information about the print statement.  You should type `q` and then hit `enter` in the text box to exit the help utility. 

Note: this exercise is pretty simple (and there's not much to put in the box)!  We just want to make sure that you have tried out this super-handy feature of Python!

In [1]:
help()


Welcome to Python 2.7!  This is the online help utility.

If this is your first time using Python, you should definitely check out
the tutorial on the Internet at http://docs.python.org/2.7/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, or topics, type "modules",
"keywords", or "topics".  Each module also comes with a one-line summary
of what it does; to list the modules whose summaries contain a given word
such as "spam", type "modules spam".

help> modules


/usr/local/lib/python2.7/dist-packages/IPython/kernel/__init__.py:13: ShimWarning: The `IPython.kernel` package has been deprecated. You should import from ipykernel or jupyter_client instead.
  "You should import from ipykernel or jupyter_client instead.", ShimWarning)
/usr/lib/python2.7/dist-packages/gobject/constants.py:24: Warning: g_boxed_type_register_static: assertion 'g_type_from_name (name) == 0' failed
  import gobject._gobject



Please wait a moment while I gather a list of all available modules...

ANSI                apt                 htmlentitydefs      rmagic
BaseHTTPServer      apt_inst            htmllib             robotparser
Bastion             apt_pkg             httplib             runpy
CDROM               aptdaemon           httplib2            samba
CGIHTTPServer       aptsources          ibus                scanext
Canvas              argparse            ihooks              sched
CommandNotFound     array               imaplib             screen
ConfigParser        ast                 imghdr              select
Cookie              asynchat            imp                 serial
Crypto              asyncore            importlib           sessioninstaller
DLFCN               atexit              imputil             sets
Dialog              atk                 indicator_keyboard  setuptools
DocXMLRPCServer     audiodev            inspect             sgmllib
FSM                 audioop             

### Exercise 4  

Start the Python interpreter and use it as a calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *.

If you run a 10 kilometer race in 43 minutes 30 seconds, what is your average time per mile? What is your average speed in miles per hour? (Hint: there are 1.61 kilometers in a mile). 

In [3]:
10/1.61

6.211180124223602

In [4]:
43.5/6.211180124223602

7.003500000000001

Average Time per Mile = 7 mins

In [10]:
60.0/7.0

8.571428571428571

Average speed in miles per hour = 8.571

## Chapter 2

#### 2.1
1. Values are letters or numbers that variables can be assigned to
2. Numbers represented in a floating-point have a decimal
3. Anything between quotation marks are strings

#### 2.2
1. A variable is an assignment statement that assigns a value to something
2. The type of a variable is the type of value it refers to such as float, boolean, or integer

#### 2.4
1. Floor division cuts of the decimal points and occurs when both of the operands are integers
2. The operateor // performs floor division

#### 2.8
1. Operators concatenate strings together
2. The * operator repeats the string multiple times

#### 2.9
1. Comments in python are starte with the # sign
2. Comments help make the code more understandable

### Exercise 2  

Assume that we execute the following assignment statements:

width = 17
height = 12.0
delimiter = '.'

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

1. width/2 = 8
2. width/2.0 = 8.0
3. heigth/3 = 4.0
4. 1 + 2 * 5 = 11
5. delimiter * 5 = '.....'

### Exercise 3  

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

    


The volume of a sphere with radius r is 4/3 $\pi r^3$. What is the volume of a sphere with radius 5? Hint: 392.7 is wrong!

In [6]:
(4.0/3.0)*3.14*125.0

523.3333333333334

Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [7]:
24.95 - (24.95*0.4)

14.969999999999999

In [9]:
(12.97 + 3) + (.75*60)

60.97

If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [11]:
(8.15)+(3*7.12)+(8.15)

37.66

I will get back home 38:06 mins after 6:52.

Answer = 7:29:06 am

## Chapter 3

#### 3.1
1. Functions are a sequence of statements that can be called and executed
2. Functions typically return a value or execute code

#### 3.2 - 3.3
1. Functions can be used for a variety of things such as converting values or other math
2. Modules are a file that containg a bunch of related functions
3. To access a function use dot notation (module.function)

##### 3.5
1. First line of a function is called the header and the rest is th body
2. Header has to end with a colon
3. Syntax for calling functions is name()

#### 3.7
1. Execution always begins at the first statement of the program
2. Executed one at a time in order from top to bottom
3. When program gets to a function it will go into the function script and run each line one at a time

#### 3.8
1. Parameters are values that are passed into a function
2. Parameters are local to the functions they are in and can be used as a variable within the function
3. Arguments are the values we send into functions as parameters

#### 3.11
1. Fruitful functions are functions that return a value
2. Void functions dont return a value
3. When calling fruitful functions you should be doing something that requires a value
4. Void functions could display something on the screen or assign values to variables

#### 3.13
1. using the from with import helps when there are multiple variables with the same name within the same module
2. syntax for from is:  from math import pi

### Exercise 3

Python provides a built-in function called len that returns the length of a string, so the value of len('allen') is 5.
Write a function named right_justify that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('allen')
                                                                 allen```
                                                                 

#### Answer

def right_justify(name) 

    spaces = 70 
    
    length = len(name) 
    
    print(' '*(spaces-length)+name) 
    
right_justify('allen')

### Exercise 5

This exercise can be done using only the statements and other features we have learned so far.

(a) Write a function that draws a grid like the following:
```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
**Hint:** to print more than one value on a line, you can print a comma-separated sequence:
print '+', '-'
If the sequence ends with a comma, Python leaves the line unfinished, so the value printed next appears on the same line.
print '+', 
print '-'
The output of these statements is '+ -'.
A print statement all by itself ends the current line and goes to the next line.

In [14]:
def cool_box():
    print('+ - - - - + - - - - +')
    print('|         |         |')
    print('|         |         |')
    print('|         |         |')
    print('|         |         |')
    print('+ - - - - + - - - - +')
    print('|         |         |')
    print('|         |         |')
    print('|         |         |')
    print('|         |         |')
    print('+ - - - - + - - - - +')
    
cool_box()

+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +


In [24]:
def draw_end(x,y):
    String1 = ' -' * (y//2)
    String2 = '+' + String1
    String3 = String2 * x,'+'
    print String3
def draw_line(x,y):
    
    
    
    print'|         '*x, '|'
    print'|         '*x, '|'
    print'|         '*x, '|'
    print'|         '*x, '|'
    print'|         '*x, '|'
def draw_Grid(a,b):
    for x in range(0,a):
        draw_end(b)
        draw_line(b)
    draw_end(b)
           
draw_Grid(3,6)
     

+ - - - - + - - - - + - - - - + - - - - + - - - - + - - - -  +
|         |         |         |         |         |          |
|         |         |         |         |         |          |
|         |         |         |         |         |          |
|         |         |         |         |         |          |
|         |         |         |         |         |          |
+ - - - - + - - - - + - - - - + - - - - + - - - - + - - - -  +
|         |         |         |         |         |          |
|         |         |         |         |         |          |
|         |         |         |         |         |          |
|         |         |         |         |         |          |
|         |         |         |         |         |          |
+ - - - - + - - - - + - - - - + - - - - + - - - - + - - - -  +
|         |         |         |         |         |          |
|         |         |         |         |         |          |
|         |         |         |         |         |    

(b) Write a function that draws a similar grid with four rows and four columns.

In [17]:
def cool_box():
    print('+ - - - - + - - - - + - - - - + - - - - +')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('+ - - - - + - - - - + - - - - + - - - - +')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('+ - - - - + - - - - + - - - - + - - - - +')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('+ - - - - + - - - - + - - - - + - - - - +')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('|         |         |         |         |')
    print('+ - - - - + - - - - + - - - - + - - - - +')
    
cool_box()

+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +


## Chapter 5.1 - 5.7

#### 5.1
1. the modulus operator yields the remaider of the first operand divided by the second
2. Operator is the %

#### 5.2
1. boolean values are either true or false
2. uses the operator ==
3. produces true if the two operands are equal and false if not
4. relational operators can be used for programming using logic

#### 5.3 - 5.5
1. Logical operators include <, >, ==
2. Conditional statements allow the program to take certain paths depending on the current state of the program
3. uses the if, else if, and else statement with logical operators

#### 5.6 - 5.7
1. Chained conditionals use the elseif and else commands
2. Each condition is checked in order and executes the first one that is true
3. nested conditionals will only execute if the one above them is completed

### Exercise 3  

Fermat’s Last Theorem says that there are no positive integers a, b, and c such that $a^n + b^n = c^n$ for any values of n greater than 2.

(a) Write a function named `check_fermat` that takes four parameters—a, b, c and n—and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$ the program should print, "Holy smokes, Fermat was wrong!" Otherwise the program should print, "No, that doesn’t work."

In [21]:
def check_fermat(a, b, c, n):
    if ((a**n + b**n) == (c**n)):
        print('Holy smokes, Fermat was wrong!')
    else:
        print('No, that does not work')

(b) Write a function that prompts the user to input values for a, b, c and n, converts them to integers, and uses check_fermat to check whether they violate Fermat’s theorem.

In [23]:
def convert_inputs(a, b, c, n):
    a_input = int(a)
    b_input = int(b)
    c_input = int(c)
    n_input = int(n)
    check_fermat(a_input, b_input, c_input, n_input)   
    
convert_inputs('1', '2', '3', '4')    

No, that does not work


### Exercise 4  

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

(a) Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [35]:
def is_triangle(a, b, c):
    if (a+b < c):
        print('No')
    elif (a+c < b):
        print('No')
    elif (b+c < a):
        print('No')
    else:
        print('Yes')
        
is_triangle(1,1,12)

No


(b) Write a function that prompts the user to input three stick lengths, converts them to integers, and uses is_triangle to check whether sticks with the given lengths can form a triangle.

In [36]:
def convert_inputs(a, b, c):
    a_input = int(a)
    b_input = int(b)
    c_input = int(c)
    is_triangle(a_input, b_input, c_input) 
    
convert_inputs('1','2','3')

Yes


## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.